tests/manual/events/KeyboardTest.java

```diff
@@ -0,0 +1,812 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.util.ArrayList;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.Timer;
+import java.util.TimerTask;
+import javafx.application.Application;
+import javafx.application.Platform;
+import javafx.event.EventHandler;
+import javafx.geometry.Insets;
+import javafx.scene.Node;
+import javafx.scene.control.Button;
+import javafx.scene.control.ChoiceBox;
+import javafx.scene.control.TextArea;
+import javafx.scene.input.KeyCharacterCombination;
+import javafx.scene.input.KeyCode;
+import javafx.scene.input.KeyEvent;
+import javafx.scene.layout.HBox;
+import javafx.scene.layout.Priority;
+import javafx.scene.layout.VBox;
+import javafx.scene.robot.Robot;
+import javafx.scene.Scene;
+import javafx.stage.Stage;
+
+/*
+ * This application tests key event handling in JavaFX. Each test uses a Robot
+ * to send platform key events and then verifies that the correct JavaFX
+ * KeyEvents are generated.
+ *
+ * To provide thorough coverage a test has to be targeted at a specific layout.
+ * Currently there are tests for U.S. (QWERTY), French (AZERTY), and German
+ * (QWERTZ) on Mac, Windows, and Linux. Since there's no way for JavaFX to force
+ * the layout or verify which layout is currently active it is up to the tester
+ * to configure the correct layout before running the test.
+ *
+ * Each language-specific test must be run against the default layout for that
+ * language. For example, the German test is designed to work with the layout
+ * labeled "German" on a Mac, not "German - Standard" or any other variant.
+ *
+ * There is also a generic test for Latin layouts which verifies that KeyCodes A
+ * through Z are reachable and generate the letters 'a' through 'z'. An even
+ * more generic test is available for non-Latin layouts which verifies that
+ * KeyCodes A through Z generate characters.
+ *
+ * None of these tests cover the top-row function keys or the Caps Lock key.
+ * They also do not cover dead keys or keys which generate accented characters
+ * (the latter don't have KeyCodes so the Robot cannot access them).
+ *
+ * These tests always check that the given KeyCode generates the expected
+ * character (if any). They can optionally check that KeyCharacterCombinations
+ * match for characters on that key. This option is disabled by default since
+ * KeyCharacterCombinations don't work reliably on most platforms (for now).
+ */
+
+public class KeyboardTest extends Application {
```

Comment:
Could you clarify how to run the test and what one should expect?

In my case, I'm running from Mac, with a Spanish physical keyboard, but changed to German keyboard from Keyboard -> Input sources -> German, and whether the Keyboard viewer is visible or not, I get:
```
[Mac] Testing 78 keys on German without combinations
Failed: code Numpad 0 did not produce any events
...
Failed: code Decimal did not produce any events
Tested 78 keys with 76 failures
```
Is that expected? Is that test only meant for a German keyboard?

Reply:
@jperedadnr You set up the test correctly. I think you need to need to allow the 'terminal' application to use the accessibility API so the Java robot can send keyboard events. I can't provide you with the exact setting since I'm away from my Mac until the end of this week. When I get back I'll send details and add a comment to the test file. Apologies, I set up my Mac so long ago I forgot about this hiccup.

Reply:
Great, that helps: after enabling Dock -> Privacy&Security -> Accessibility -> Terminal, tests for all 5 keyboards work, but only "without combinations":
```
[Mac] Testing 83 keys on U.S. English without combinations
Tested 83 keys with 0 failures
...
[Mac] Testing 78 keys on German without combinations
Tested 78 keys with 0 failures
```
However, without Keypad/All combinations fail:
```
[Mac] Testing 80 keys on French with all combinations
Failed: code A did not match combination 'A'
...
Failed: code Minus did not match combination '_'
Tested 80 keys with 45 failures
``` 
is that expected?

Reply:
Yes, the KeyCharacterCombination failures are expected. Fixing those will require a separate PR. I’m hoping Linux PR #694 will be accepted and then I can extend that approach to the Mac.